### PR TITLE
adding docstrings to methods/functions

### PR DESCRIPTION
**Pull request by mwinton**
Hi Ram,

I had some spare time during 205 class, so I've gone through most of our files and added docstrings. Please review and let me know when I got something wrong.  My biggest questions are around the image feature_idx, and the Image.load() method.

If you have time, could you go through the files under `vqa/model` and document them similarly?

### CODE REVIEW

File: bin/visualqa.py [hunk 1]
```diff
@@ -318,7 +336,19 @@ def plot_train_metrics(train_stats, options, plot_type='epochs'):
 
 
 def train(model, dataset, options, val_dataset=None, attention_model=None):
-
+    """
+        Trains the Keras model.
+        
+        Args:
+            model = a Keras Model instance
+            dataset = VQADatasete instance to train on
```
Comment (ram-iyer) [inline bin/visualqa.py]: just a typo

File: bin/visualqa.py [hunk 2]
```diff
@@ -318,7 +336,19 @@ def plot_train_metrics(train_stats, options, plot_type='epochs'):
 
 
 def train(model, dataset, options, val_dataset=None, attention_model=None):
-
+    """
+        Trains the Keras model.
```
Comment (ram-iyer) [inline bin/visualqa.py]: .. and saves results to disk

File: bin/visualqa.py [hunk 3]
```diff
@@ -409,6 +439,17 @@ def train(model, dataset, options, val_dataset=None, attention_model=None):
         val_dataset.dataset_type = DatasetType.VALIDATION
 
 def test(model, dataset, options, attention_model=None):
+    """
+        Run predictions using a pre-trained Keras model.
```
Comment (ram-iyer) [inline bin/visualqa.py]: ..and saves results to disk

File: bin/visualqa.py (hunk 2 above)
Comment (ram-iyer) [inline bin/visualqa.py]: we should also mention that it saves the models and metrics

File: bin/visualqa.py [hunk 4]
```diff
@@ -42,8 +42,13 @@
 
 # ------------------------------- SCRIPT FUNCTIONALITY -------------------------------
 
-def main(options):
-
+def main(options):    
+    """
+        Main entry point for training and running models.
```
[the remark this answers is not in the document]
Comment (mwinton) [reply bin/visualqa.py]: done

File: bin/visualqa.py [hunk 5]
```diff
@@ -129,15 +134,22 @@ def main(options):
...
-    
+def load_dataset(dataset_type, options, answer_one_hot_mapping=None, tokenizer=None):    
     """
         Load the dataset from disk if available. If not, build it from the questions/answers json and image embeddings
         If this is the training dataset, retrieve the answer one hot mapping from disk or re-create it.
+        
+        Args:
+            dataset_type = DatasetType.TRAIN or DatasetType.TEST (integer constants)
+            options = ModelOptions object containing parameters
+            answer_one_hot_mapping = dictionary mapping words to OHE index (optional)
```
[the remark this answers is not in the document]
Comment (mwinton) [reply bin/visualqa.py]: done

File: bin/visualqa.py [hunk 6]
```diff
@@ -129,15 +134,22 @@ def main(options):
...
 
-def load_dataset(dataset_type, options, answer_one_hot_mapping=None, tokenizer=None):
-    
+def load_dataset(dataset_type, options, answer_one_hot_mapping=None, tokenizer=None):    
     """
         Load the dataset from disk if available. If not, build it from the questions/answers json and image embeddings
         If this is the training dataset, retrieve the answer one hot mapping from disk or re-create it.
+        
+        Args:
+            dataset_type = DatasetType.TRAIN or DatasetType.TEST (integer constants)
```
[the remark this answers is not in the document]
Comment (mwinton) [reply bin/visualqa.py]: done

File: bin/visualqa.py [hunk 7]
```diff
@@ -258,7 +270,13 @@ def plot_train_metrics(train_stats, options, plot_type='epochs'):
         Generate and save figure with plot of train and validation losses.
         Currently, plot_type='epochs' is the only option supported.
         
-        train_stats is a Keras History object.  History.history is a dict containing lists
+        Args:
+            train_stats = Keras History instance; History.history is a dict containing lists
```
[the remark this answers is not in the document]
Comment (mwinton) [reply bin/visualqa.py]: done

File: bin/visualqa.py (hunk 3 above)
Comment (mwinton) [reply bin/visualqa.py]: done

File: bin/visualqa.py (hunk 2 above)
Comment (mwinton) [reply bin/visualqa.py]: done

File: bin/visualqa.py (hunk 1 above)
Comment (mwinton) [reply bin/visualqa.py]: done

**Pull request merged**